checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java

```diff
@@ -426,39 +428,44 @@ private void trackInvocationResult(Set<Obligation> obligations, Node node) {
+    if (mustCallAliases.isEmpty()
         && node instanceof MethodInvocationNode
         && typeFactory.returnsThis((MethodInvocationTree) tree)) {
-      mustCallAlias =
-          removeCastsAndGetTmpVarIfPresent(((MethodInvocationNode) node).getTarget().getReceiver());
+      mustCallAliases.add(
+          removeCastsAndGetTmpVarIfPresent(
+              ((MethodInvocationNode) node).getTarget().getReceiver()));
     }
 
     ResourceAlias tmpVarAsResourceAlias = new ResourceAlias(new LocalVariable(tmpVar), tree);
-    if (mustCallAlias instanceof FieldAccessNode) {
-      // Do not track the call result if the MustCallAlias argument is a field.  Handling of
-      // @Owning fields is a completely separate check, and there is never a need to track an alias
-      // of
-      // a non-@Owning field, as by definition such a field does not have obligations!
-    } else if (mustCallAlias instanceof LocalVariableNode) {
-      Obligation obligationContainingMustCallAlias =
-          getObligationForVar(obligations, (LocalVariableNode) mustCallAlias);
-      // If mustCallAlias is a local variable already being tracked, add tmpVarAsResourceAlias
-      // to the set containing mustCallAlias.
-      if (obligationContainingMustCallAlias != null) {
-        Set<ResourceAlias> newResourceAliasSet =
-            FluentIterable.from(obligationContainingMustCallAlias.resourceAliases)
-                .append(tmpVarAsResourceAlias)
-                .toSet();
-        obligations.remove(obligationContainingMustCallAlias);
-        obligations.add(new Obligation(newResourceAliasSet));
+    for (Node mustCallAlias : mustCallAliases) {
+      if (mustCallAlias instanceof FieldAccessNode) {
+        // Do not track the call result if the MustCallAlias argument is a field.  Handling of
+        // @Owning fields is a completely separate check, and there is never a need to track an
+        // alias
+        // of
+        // a non-@Owning field, as by definition such a field does not have obligations!
+      } else if (mustCallAlias instanceof LocalVariableNode) {
+        Obligation obligationContainingMustCallAlias =
+            getObligationForVar(obligations, (LocalVariableNode) mustCallAlias);
+        // If mustCallAlias is a local variable already being tracked, add tmpVarAsResourceAlias
+        // to the set containing mustCallAlias.
+        if (obligationContainingMustCallAlias != null) {
+          Set<ResourceAlias> newResourceAliasSet =
+              FluentIterable.from(obligationContainingMustCallAlias.resourceAliases)
+                  .append(tmpVarAsResourceAlias)
+                  .toSet();
+          obligations.remove(obligationContainingMustCallAlias);
+          obligations.add(new Obligation(newResourceAliasSet));
+        }
       }
-    } else {
-      // If mustCallAlias is neither a field nor a local already in the set of obligations,
-      // add it to a new set.
+    }
+    if (mustCallAliases.isEmpty()) {
```

Comment:
I don't think this logic is equivalent to what was present before.  Before, if `mustCallAlias` was neither a `FieldAccessNode` nor a `LocalVariableNode`, then `tmpVarAsResourceAlias` would be added to a new obligation set.  Now, if `mustCallAliases` is a singleton list where the contained node is neither a `FieldAccessNode` nor a `LocalVariableNode`, nothing happens.  Can we figure out which logic is correct, and add a test case to expose the difference?

Reply:
In the previous logic `mustCallAlias` could be null(it's equivalent to empty list now) and else branch handled these cases. I can't think of a test case that `mustCallAlias`  is neither a `FieldAccessNode` nor a `LocalVariableNode` after line 437 where we replace it with its tempVar.

Reply:
I think you're right.  I'll add another comment on how to make this more clear.

Reply:
Sorry, one last comment here (that I think I wrote before but somehow got lost).  Can we move this `if` check at line 466 above the previous `for (Node mustCallAlias : mustCallAliases) {` loop, and then put the loop in an `else` block for the `if`?  That will make the logic clearer.